### DISCUSSION TITLE
why my bpftrace program can not run on linux kernel version 4.18

### DISCUSSION THREAD

**Discussion by jiufei19** [bpftrace/bpftrace#2661] [Q&A] [answered]
```
/* FILENAME: vfs_open.bt */

#include <linux/path.h>
#include <linux/dcache.h>

kprobe:vfs_open
{
    printf("open path: %s\n", str(((struct path *)arg0)->dentry->d_name.name));
}

```
Since struct path and struct dentry are defined in the path.h and dcache.h in the directory of `/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/include`, I run this program on linux kernel version 4.18 in this way:
```
sudo bpftrace -I /usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/include/ -I /usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/arch/x86/include vfs_open.bt
```
But bpftrace always reports a lot of errors:

```
...
/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/arch/x86/include/asm/atomic64_64.h:178:42: error: unknown type name 'atomic64_t'
/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/arch/x86/include/asm/atomic64_64.h:180:9: error: returning 'void' from a function with incompatible result type 'long'
fatal error: too many errors emitted, stopping now [-ferror-limit=]
 "ERROR: Include headers with missing type definitions or install BTF information to your system."
```
Of course on kernel ver 4.18, there is no BTF support exists, but I have given the right include dir search path as above, why bpftrace still reports error? 

I have checked all include file directories, but I still can not make it, oops!

anyone can help me?

**Comment by viktormalik** [comment]
Did you try to run it without the explicit include directories? As long as `uname -r` returns the current version number, bpftrace should be able to handle it itself.

On a different note, looking at your kernel version, that looks like a reasonably new RHEL (CentOS) which should definitely have BTF so you can even try to remove the includes and see if it works.

**Reply by jiufei19** [reply]
thanks， I know when BTF exists, no need to specify include search path. For example, when I run this code under kernel 6.2.9, it works without any search path specified.

when under kernel 4.18，without explicit include dir, then bpftrace will report:

definitions.h:3:10: fatal error: 'linux/path.h' file not found

**Reply by viktormalik** [reply]
Could you try to remove the `include ...` lines from the script and run it like that? They prevent bpftrace from pulling necessary information from BTF.

**Reply by jiufei19** [reply]
well, kernel v4.18 doesn't support BTF, so if I remove #include lines from the script, errors are as follow:
[admin@node1 bpf]$ sudo bpftrace vfs_open.bt 
vfs_open.bt:8:31-52: ERROR: Unknown struct/union: 'struct path'
    printf("open path: %s\n", str(((struct path *)arg0)->dentry->d_name.name));
vfs_open.bt:8:31-79: ERROR: str() expects an integer or a pointer type as first argument (none provided)
    printf("open path: %s\n", str(((struct path *)arg0)->dentry->d_name.name));

**Reply by viktormalik** [reply]
Could you please paste the output of `bpftrace --info`?

**Comment by jiufei19** [comment]
thanks a lot, here is the output:
-------------------------------------
***@***.*** bpf]$ sudo bpftrace --info
[sudo] admin 的密码：
System
OS: Linux 4.18.0-80.el8.x86_64 #1 SMP Tue Jun 4 09:19:46 UTC 2019
Arch: x86_64


Build
version: v0.12.1
LLVM: 12.0.0
ORC: v2
foreach_sym: yes
unsafe uprobe: no
bfd: yes
bpf_attach_kfunc: yes
bcc_usdt_addsem: yes
bcc bpf_attach_uprobe refcount: yes
libbpf: yes
libbpf btf dump: yes
libbpf btf dump type decl: yes


Kernel helpers
probe_read: yes
probe_read_str: yes
probe_read_user: no
probe_read_user_str: no
probe_read_kernel: no
probe_read_kernel_str: no
get_current_cgroup_id: yes
send_signal: no
override_return: no
get_boot_ns: no
dpath: no


Kernel features
Instruction limit: 131072
Loop support: no
btf (depends on Build:libbpf): no
map batch (depends on Build:libbpf): no
uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): no


Map types
hash: yes
percpu hash: yes
array: yes
percpu array: yes
stack_trace: yes
perf_event_array: yes


Probe types
kprobe: yes
tracepoint: yes
perf_event: yes
kfunc: no
iter:task: no
iter:task_file: no


***@***.*** bpf]$



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "iovisor/bpftrace"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年7月4日(星期二) 下午4:28
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [iovisor/bpftrace] why my bpftrace program can not run on linux kernel version 4.18 (Discussion #2661)


 
Could you please paste the output of bpftrace --info?

**Reply by viktormalik** [reply, marked as answer]
You're using bpftrace 0.12.1 which is fairly (over 2 years) old. IIRC, it doesn't have the ability to completely pull type definitions from BTF. I'd suggest upgrading to a newer bpftrace (you'll probably need to build it from source).

Still, your script should work with correct headers included without the need to specify explicit include paths. It seems that bpftrace cannot properly find your kernel source directory. You could try to pass it directly via env variable:
```
BPFTRACE_KERNEL_SOURCE=/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/ bpftrace ...
```

**Comment by jiufei19** [comment]
it works with the env varible set.
thanks a lot!

Yes, when I set the env varible as follow:

[admin@node1 bpf]$ sudo BPFTRACE_KERNEL_SOURCE=/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/ bpftrace vfs_open.bt
Attaching 1 probe...
open path: system.journal
open path: imjournal.state.tmp
open path: comm
open path: comm
...

------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "iovisor/bpftrace"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年7月4日(星期二) 下午4:42
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [iovisor/bpftrace] why my bpftrace program can not run on linux kernel version 4.18 (Discussion #2661)





 
You're using bpftrace 0.12.1 which is fairly (over 2 years) old. IIRC, it doesn't have the ability to completely pull type definitions from BTF. I'd suggest upgrading to a newer bpftrace (you'll probably need to build it from source).
 
Still, your script should work with correct headers included without the need to specify explicit include paths. It seems that bpftrace cannot properly find your kernel source directory. You could try to pass it directly via env variable:
 BPFTRACE_KERNEL_SOURCE=/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/ bpftrace ...

**Comment by jiufei19** [comment]
well, I am still a little wondering that when I set the env variable as your suggestion, and I didn't specify any include search path,&nbsp; why bpftrace can still find the linux/path.h and linux/dcache.h without BTF support?




为下一代
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "iovisor/bpftrace"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年7月4日(星期二) 下午4:42
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [iovisor/bpftrace] why my bpftrace program can not run on linux kernel version 4.18 (Discussion #2661)





 
You're using bpftrace 0.12.1 which is fairly (over 2 years) old. IIRC, it doesn't have the ability to completely pull type definitions from BTF. I'd suggest upgrading to a newer bpftrace (you'll probably need to build it from source).
 
Still, your script should work with correct headers included without the need to specify explicit include paths. It seems that bpftrace cannot properly find your kernel source directory. You could try to pass it directly via env variable:
 BPFTRACE_KERNEL_SOURCE=/usr/src/kernels/4.18.0-348.7.1.el8_5.x86_64/ bpftrace ...

**Reply by viktormalik** [reply]
This is b/c bpftrace automatically adds multiple include paths based on that dir. See [utils.cpp:414](https://github.com/iovisor/bpftrace/blob/master/src/utils.cpp#L414-L422), `ksrc` is obtained from the env variable.

**Comment by fbs** [comment]
Please use codeblocks, using 3 backticks

````
```
mycode
```
````

it really improves readability

**Reply by jiufei19** [reply]
thanks for your suggestion.
